### PR TITLE
feat(config): migrate inline with the same sort

### PR DESCRIPTION
**Pull request by pret-a-porter**
## Changes:

Refactor migrations to provide ability to migrate and save original order of properties in configuration

## Context:

#11459

## Documentation

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
OK please get one step working first to demonstrate. I was hoping you would use the code I posted earlier which solves the majority of the problem in one go

**Comment by HonkingGoose** [comment]
From our Pull Request template:

> Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. 
> 
> This makes it harder for us to review your work because we don't know what has changed. 
> 
> PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch.

**Comment by rarkins** [comment]
@pret-a-porter could you summarize how much you've implemented, how it can be tested/verified, etc? e.g. I see one test which compares Object.keys(), is that it? It's the sorting of keys which is most important, so standard jest equivalency tests won't properly verify because they sort keys before comparing.

File: lib/config/migration.ts [hunk 1]
```diff
@@ -50,7 +33,8 @@ export function migrateConfig(
         optionTypes[option.name] = option.type;
       });
     }
-    const migratedConfig = clone(config) as MigratedRenovateConfig;
+    const newConfig = MigrationsService.run(config);
+    const migratedConfig = clone(newConfig) as MigratedRenovateConfig;
```
Comment (viceice) [inline lib/config/migration.ts:37]: @rarkins shouldn't the clone be done in migration service?

File: lib/config/migrations/custom/go-mod-tidy-migration.spec.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+describe('config/migrations/custom/go-mod-tidy-migration', () => {
+  it('should add postUpdateOptions option when true', () => {
+    const migratedConfig = MigrationsService.run({
+      gomodTidy: true,
+      postUpdateOptions: ['test'],
+    });
+
+    expect(migratedConfig.gomodTidy).toBeUndefined();
+    expect(migratedConfig.postUpdateOptions).toContain('test');
+    expect(migratedConfig.postUpdateOptions).toContain('gomodTidy');
```
Comment (viceice) [inline lib/config/migrations/custom/go-mod-tidy-migration.spec.ts:12]: Isn't a `toEqual([...])` simpler and cleaner  here, as used in other tests.

File: lib/config/migrations/base/abstract-migration.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    migratedConfig: RenovateConfig
+  ) {
+    this.propertyName = propertyName;
+    this.originalConfig = originalConfig;
+    this.migratedConfig = migratedConfig;
+  }
+
+  abstract run(): void;
+
+  protected delete(property: string): void {
```
Comment (viceice) [inline lib/config/migrations/base/abstract-migration.ts:23]: What about making property optional and use `this.PropertyName` then?

This would make some migrations shorter and simpler. 🤗

File: lib/config/migrations/migrations-service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+describe('config/migrations/migrations-service', () => {
+  it('should remove deprecated properties', () => {
+    for (const property of MigrationsService.removedProperties) {
+      const originalConfig: RenovateConfig = {
+        [property]: 'test',
+      };
+
+      const migratedConfig = MigrationsService.run(originalConfig);
+      expect(migratedConfig[property]).toBeUndefined();
```
Comment (viceice) [inline lib/config/migrations/migrations-service.spec.ts:12]: I think we should use `not.toHaveProperty` matcher?

Your code will not catch properties with `undefined` value assignment.

File: lib/config/migrations/migrations-service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    for (const [
+      oldPropertyName,
+      newPropertyName,
+    ] of MigrationsService.renamedProperties.entries()) {
+      const originalConfig: RenovateConfig = {
+        [oldPropertyName]: 'test',
+      };
+
+      const migratedConfig = MigrationsService.run(originalConfig);
+      expect(migratedConfig[oldPropertyName]).toBeUndefined();
```
Comment (viceice) [inline lib/config/migrations/migrations-service.spec.ts:26]: See comment above.

File: lib/config/migration.ts (hunk 1 above)
Comment (rarkins) [reply lib/config/migration.ts:37]: Yes, I think so. @pret-a-porter any reason that won't work?

Comment (pret-a-porter) [reply lib/config/migration.ts:37]: To save order of properties, instead of clone original config we create empty object `const migratedConfig: RenovateConfig = {};` and then fill it step by step go over properties of original config

File: lib/config/migrations/base/abstract-migration.ts (hunk 3 above)
Comment (pret-a-porter) [reply lib/config/migrations/base/abstract-migration.ts:23]: I thought about it as well, but this code `this.delete()` inside migration does not look descriptive for me

**Comment by pret-a-porter** [comment]
> @pret-a-porter could you summarize how much you've implemented, how it can be tested/verified, etc? e.g. I see one test which compares Object.keys(), is that it? It's the sorting of keys which is most important, so standard jest equivalency tests won't properly verify because they sort keys before comparing.

To summarize:

1. Implemented class MigrationService, which should run migrations and save original order of properties. Basically yes, only one test checks that service save order of properties properly. I have tried to use Jest snapshots, but looks like it sorts properties before creating snapshot
2. Implemented basic migrations, to handle simple cases with removing and renaming properties in configuration
3. Implemented few custom migrations, to show how to handle more complicated migrations

File: lib/config/migration.ts (hunk 1 above)
Comment (rarkins) [reply lib/config/migration.ts:37]: Does it need cloning? And if so, can the clone be done before the `return` statement instead of after?

Comment (pret-a-porter) [reply lib/config/migration.ts:37]: Yeah, we still need cloning. Maybe when all migrations will be refactored will be possible not to use `clone`.

We need cloning, because we use `dequal` to compare original and migrated configs. https://github.com/lukeed/dequal/blob/master/src/index.js under the hood this library by default compares values by `===` (it is not very clear from the docs).

I have tried to do `clone` before statement, but in this case some unit tests failed. I did not investigate deeper, if it is important I will do.

Comment (viceice) [reply lib/config/migration.ts:37]: Ok, so leave as is for now

**Review by viceice — APPROVED** [review]

**Comment by rarkins** [comment]
@pret-a-porter can you provide some logs of comparing the migrated config log of this new code compared to the existing code? i.e. on "real" repos

**Comment by rarkins** [comment]
The whole purpose of this issue was key ordering in migrated configs but there's almost no testing of that

**Comment by pret-a-porter** [comment]
image
@rarkins I have tried to run against https://github.com/pret-a-porter-testing/renovate-reproduction-11459/blob/main/renovate.json

**Comment by rarkins** [comment]
> image @rarkins I have tried to run against [pret-a-porter-testing/renovate-reproduction-11459@`main`/renovate.json](https://github.com/pret-a-porter-testing/renovate-reproduction-11459/blob/main/renovate.json?rgh-link-date=2021-11-06T22%3A13%3A44Z)

When you test, it's best if you can include some *non* migrated config too, so that we can observe they are ordered correctly even if not migrated.

**Comment by pret-a-porter** [comment]
@rarkins sorry, I did not get it. In have created repository with non migrated config https://github.com/pret-a-porter-testing/renovate-reproduction-11459. On screenshot you can see lines with `oldConfig` and `newConfig`. Shall I do something else?

**Comment by rarkins** [comment]
> @rarkins sorry, I did not get it. I have created repository with non migrated config [pret-a-porter-testing/renovate-reproduction-11459](https://github.com/pret-a-porter-testing/renovate-reproduction-11459). On screenshot you can see lines with `oldConfig` and `newConfig`. Shall I do something else?

Yes, you need to do something else. Every config option in newConfig has been migrated. You should test with a mix of migrated and non-migrated (does not need migration) config.

**Comment by pret-a-porter** [comment]
Screenshot 2021-11-07 at 15 41 25
@rarkins done

**Comment by rarkins** [comment]
Can you put ignoreTests last and verify that it remains last?

**Comment by rarkins** [comment]
Actually, have a mix of non-migrated fields first, last and in the middle of migrated and verify ordering remains as expected

**Comment by pret-a-porter** [comment]
image
@rarkins done as well

**Comment by pret-a-porter** [comment]
@rarkins shall I do any other validations regarding this PR?

**Comment by rarkins** [comment]
> image @rarkins done as well

This test doesn't interleave migrated and non-migrated. It has all migrated first, then non-migrated. Can you e.g. move `versionScheme` last in the old config?

**Comment by pret-a-porter** [comment]
Screenshot 2021-11-13 at 21 44 06
@rarkins it works fine

**Review by rarkins — APPROVED** [review]

File: lib/config/migration.ts [hunk 6]
```diff
@@ -5,7 +5,7 @@ import { logger } from '../logger';
 import { clone } from '../util/clone';
```
Comment (HonkingGoose) [inline lib/config/migration.ts:5]: Question: does this PR fully resolve issue #11459? Or is more work needed after this PR gets merged?

Comment (pret-a-porter) [reply lib/config/migration.ts:5]: Not fully. All migrations should be refactored by the new way https://github.com/renovatebot/renovate/issues/11457. Sorting of refactored migrated properties will be saved.
